q2cli/handlers.py

```diff
@@ -0,0 +1,207 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+import os
+import collections
+
+import click
+import qiime
+
+import q2cli.util
+
+
+NoDefault = collections.namedtuple('NoDefault', [])()
+
+
+class ValueNotFoundException(Exception):
+    """Raised when a value cannot be found."""
+
+
+class Handler:
+    def __init__(self, name, prefix='', default=NoDefault):
+        self.name = name
+        self.click_name = prefix + name
+        self.default = default
+
+    @property
+    def normalized_name(self):
+        return q2cli.util.normalize_name(self.click_name)
+
+    def get_click_options(self):
+        raise NotImplementedError()
+
+    def get_value(self, arguments, fallback=None):
+        raise NotImplementedError()
+
+    def _locate_value(self, arguments, fallback, name=None, click_name=None,
+                      normalized_name=None):
+        if name is None:
+            name = self.name
+        if click_name is None:
+            click_name = self.click_name
+        if normalized_name is None:
+            normalized_name = self.normalized_name
+
+        # Is it in args?
+        v = arguments[click_name]
+        if v is not None:
+            return v
+
+        # Does our fallback know about it?
+        if fallback is not None:
+            try:
+                return fallback(name, normalized_name)
+            except ValueNotFoundException:
+                pass
+
+        # Do we have a default?
+        if self.default is not NoDefault:
+            return self.default
+
+        # Give up
+        raise ValueNotFoundException()
+
+
+class OutputDirHandler(Handler):
+    """Meta handler which returns a fallback function as its value."""
+
+    def __init__(self):
+        super().__init__('output_dir')
+
+    def get_click_options(self):
+        yield click.Option(
+            ['--' + self.normalized_name],
+            type=click.Path(exists=False, dir_okay=True, file_okay=False),
+            help='Output unspecified results to a directory')
+
+    def get_value(self, arguments, fallback=None):
+        def fallback_(name, normalized_name):
+            path = self._locate_value(arguments, fallback=fallback)
+            os.makedirs(path)
```

Comment:
This isn't quite working, it is getting called for each output, which `os.makedirs` doesn't like very much...